src/deploy-entities.ts

```diff
@@ -0,0 +1,101 @@
+ * for each entity deployed.
+ * @public
+ */
+export async function deployEntitiesFromPointerChanges(
+  components: Pick<SnapshotsFetcherComponents, 'logs' | 'metrics' | 'fetcher'> & {
+    deployer: IDeployerComponent
+  },
+  options: PointerChangesDeployedEntityStreamOptions,
+  contentServer: string,
+  shouldStopStream: () => boolean,
+  increaseLastTimestamp: (contentServer: string, ...newTimestamps: number[]) => void
+) {
+  const logger = components.logs.getLogger('deployEntitiesFromPointerChanges')
+  const deployments = getDeployedEntitiesStreamFromPointerChanges(components, options, contentServer)
+
+  for await (const deployment of deployments) {
+    // if the stream is closed then we should not process more deployments
+    if (shouldStopStream()) {
+      logger.debug('Canceling running stream')
+      return
+    }
+
+    await components.deployer.deployEntity(
+      {
+        ...deployment,
+        markAsDeployed: async function () {
+          components.metrics.increment('dcl_entities_deployments_processed_total', { source: 'pointer-changes' })
+          // update greatest processed timestamp
+          increaseLastTimestamp(contentServer, deployment.localTimestamp)
+        }
+      },
+      [contentServer]
+    )
+  }
+}
+
+/**
+ * This function streams and deploys the entities of a snapshot. When the deployer marks all the entities as deployed,
+ * it saves the snapshot as processed.
+ * @public
+ */
+export async function deployEntitiesFromSnapshot(
+  components: Pick<
+    SnapshotsFetcherComponents,
+    'metrics' | 'logs' | 'storage' | 'processedSnapshotStorage' | 'snapshotStorage'
+  > & {
+    deployer: IDeployerComponent
+  },
+  options: SnapshotDeployedEntityStreamOptions,
+  snapshotHash: string,
+  servers: Set<string>,
+  shouldStopStream: () => boolean
+) {
+  const logger = components.logs.getLogger('deployEntitiesFromSnapshot')
+  const stream = getDeployedEntitiesStreamFromSnapshot(components, options, snapshotHash, servers)
+  let snapshotWasCompletelyStreamed = false
+  let numberOfStreamedEntities = 0
+  let numberOfProcessedEntities = 0
+  async function saveIfStreamEndedAndAllEntitiesWereProcessed() {
+    if (snapshotWasCompletelyStreamed && numberOfStreamedEntities === numberOfProcessedEntities) {
+      await components.processedSnapshotStorage.saveAsProcessed(snapshotHash)
+      components.metrics.increment('dcl_processed_snapshots_total', { state: 'saved' })
+    }
+  }
+  for await (const entity of stream) {
+    if (shouldStopStream()) {
+      logger.debug('Canceling running sync snapshots stream')
+      return
+    }
+    // schedule the deployment in the deployer. the await DOES NOT mean that the entity was deployed entirely
+    // if the deployer is not synchronous. For example, the batchDeployer used in the catalyst just add it in a queue.
+    // Once the entity is truly deployed, it should call the method 'markAsDeployed'
+    await components.deployer.deployEntity(
```

Comment:
if this operation is awaited:
- do we really leverage the parallelism in the queue?
- if not, do we really need to return a promise? since it already has a callback, mixing both approaches may lead to confusion while maintaining the lib

```suggestion
    components.deployer.deployEntity(
```

Reply:
🤔 I think we do. `deployEntity` does not mean the entity was deployed, but the batchDeployer makes some async calls before adding the job to the queue.
Not sure if I completely follow it, do you suggest changes in the firm of `deployEntity` ?

[Batch deployer](https://github.com/decentraland/catalyst/blob/04a18d9cf0e3ebf3147713ffe34fc621968c2aa6/content/src/service/synchronization/batchDeployer.ts#L19)

Reply:
Ok, I took a read to the procedure and it _must_ be awaited. If we are making a breaking change in the api though, I'd rename it to "scheduleEntityDeployment" so it can be awaited and that way we provide enough information to the reader

Reply:
Well, theoretically we're not making a breaking change as the name and the functionality of the batch deployer `deployEntity` function has been there for a long time ago. But I like the change of the name to `scheduleEntityDeployment`, will do.

Reply:
Renamed in commit ([0544e47](https://github.com/decentraland/snapshots-fetcher/pull/106/commits/0544e47a90bae22bb7a50bca11d94a341ad5a710))